### DISCUSSION TITLE
Integration with standard github code review processes

### DISCUSSION THREAD

**Discussion by slymon99** [arxanas/git-branchless#716] [Q&A] [answered]
Hello,

I really love this project - after moving on from [big tech company], I really missed the version control tooling there and this does a great job of bringing a lot of those features to git.

I'm wondering what other people are doing to integrate with github for code reviews in this tool? My workflow is currently:

1. Do my work in stacks of commits
2. Add branches to each commit (e.g. somefeature_0, somefeature_1, somefeature_2)
3. Move those branches with git branch -f if the underlying commit changes
4. Create PR's per branch, each merging into the prior branch (0 -> main, 1 -> 0, 2 -> 1).
5. Manually add comments to each PR linking to the prior / next commit (stacked diffs)
6. Once approved, merge in reverse order (first 2 into 1, then (1+2) into 0, then (0+1+2) into main)

The main things I miss from [big tech company] is the automatic syncing between my stack and the review tool (e.g. automatically have a stack of PRs, one per commit, with forward/backlinks), as well as having all the PR-related information *on the commit*, so when you are moving commits around, they keep all the metadata.

A simple solution here might be just having ephemeral branches per commit, and then have the ability to generate a stack of PRs from a stack? This way you can keep all the relevant information (the PR comments, etc) on the *branch*, but the branch is really 1:1 with a commit on the stack. Has anyone implemented any scripting to do anything like this? It would have to integrate with some of the auto-rebase tools in here e.g. restack/sync - when you restack and create new commits, you have the move the branches as well.

Interested to see what other people's github workflows are.

Thanks!

**Comment by arxanas** [comment, marked as answer]
There's some discussion in these threads:  https://github.com/arxanas/git-branchless/discussions/45, https://github.com/arxanas/git-branchless/discussions/47

> Move those branches with git branch -f if the underlying commit changes

You're probably already aware, but for the case of updating existing commits, git-branchless will automatically move branches to point to the newest versions of the commits (or, if it can't, it will prompt you to run `git restack`).

> Once approved, merge in reverse order (first 2 into 1, then (1+2) into 0, then (0+1+2) into main)

If you have push access and are pushing branches directly to that repository and opening PRs using those, you can set the "base" branch in GitHub to the previous branch, and it will let you merge 0 first and it will automatically update descendant branches.

You can also try a tool like https://graphite.dev/ as a web interface to review and merge stacked PRs. I think it has a merge-all action somewhere, and it should also be better at tracking comments for previous versions of the PR.

Another alternative is to just do multiple commits per PR and use GitHub's tooling to review and merge them. It's not great, but it has gotten better over the last few years.

> Has anyone implemented any scripting to do anything like this?

git-branchless's `git submit` command would ideally do this do this. Currently, it just manages force-pushing the stack of branches, but not creating PR objects on the remote Git hosting provider. There's some discussion on the possible interace here: https://github.com/arxanas/git-branchless/discussions/564. Feel free to post there or in the [Discord chat](https://discord.gg/caYQBJ82A4) if you want to contribute to this area.

There's a lot of existing GitHub PR management tools. You might also want to consider one of these: https://github.com/gitext-rs/git-stack/blob/main/docs/comparison.md

GitLab lets you create PRs on push: https://github.com/arxanas/git-branchless/discussions/45#discussioncomment-3364792, so it's unfortunate that GitHub doesn't.

Here's an example script to run a command on each branch with the context of its previous branch, from @mlcui-google in https://github.com/arxanas/git-branchless/discussions/624#discussioncomment-4406634. Primarily I'm showing it because it has an example of a complicated revset in case you need to do something special:

```bash
for branch in $(git query --branches 'draft () & branches()'); do
  previous_branch=$(git query --branches "only(branches() & ancestors($branch^), (branches() & ancestors($branch^))^)")
  git branch --set-upstream-to="$previous_branch" "$branch"
done
```

Other ideas are @allen-munsch's scripts here: https://github.com/arxanas/git-branchless/discussions/45#discussioncomment-3509548 or @pokey's scripts here: https://github.com/arxanas/git-branchless/discussions/45#discussioncomment-3724425.

You could do something pretty similar to iterate over commits and 1) create a new branch with some arbitrary name, and 2) run e.g. `gh pr create --base <previous-branch> ...` to create the entire stack of PRs. Then you can use `git submit` to update it later.

> A simple solution here might be just having ephemeral branches per commit

[Jujutsu](https://github.com/martinvonz/jj) can create ephemeral branches for `jj git push` (if I understand correctly), although it doesn't have any particular GitHub integration, either.

**Reply by slymon99** [reply]
Really in depth answer, thank you. I hadn't seen git submit - that covers a significant amount of the annoyance I faced. I hadn't seen git submit on any of the docs, I see the discussions since its a newer command.

**Reply by mlcui-corp** [reply]
> You can also try a tool like https://graphite.dev/ as a web interface to review and merge stacked PRs. I think it has a merge-all action somewhere, and it should also be better at tracking comments for previous versions of the PR.

Meta also has [ReviewStack](https://sapling-scm.com/docs/addons/reviewstack), which seems to be intended for Sapling users but also supports [ghstack](https://github.com/ezyang/ghstack) which should work with git-branchless (although I have not tried it out). This could give a better experience than GitHub's review tool, as it doesn't handle force pushes well.
